File: server/src/oracle/yahoo.ts
/**
 * Yahoo Finance API Client
 *
 * Provides:
 * - ISIN to ticker symbol resolution
 * - Price fetching (current)
 * - Historical price fetching (daily candles)
 * - FX rate fetching
 * - Dividend event fetching
 *
 * All requests go through a global rate limiter to avoid hitting Yahoo's limits.
 */

import { YAHOO_SEARCH_URL, YAHOO_CHART_URL } from '../config';
import { yahooRateLimiter } from './rateLimiter';

// =============================================================================
// Types
// =============================================================================

export interface YahooPriceInfo {
	price: number;
	currency: string;
	instrumentType: string;
	timestamp: Date;
}

export interface YahooSearchResult {
	symbol: string;
	exchange?: string;
	exchangeDisplay?: string;
	quoteType?: string;
	shortname?: string;
}

export interface YahooCandle {
	date: string; // YYYY-MM-DD
	open: number;
	high: number;
	low: number;
	close: number;
	adjClose: number;
	volume: number;
}

export interface YahooDividend {
	exDate: string; // YYYY-MM-DD
	amount: number;
}

export interface YahooHistoricalData {
	candles: YahooCandle[];
	dividends: YahooDividend[];
	currency: string;
}

export interface YahooIntradayCandle {
	timestamp: number; // Unix timestamp in seconds
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface YahooIntradayData {
	candles: YahooIntradayCandle[];
	currency: string;
}

export class YahooFinanceError extends Error {
	constructor(
		message: string,
		public readonly code: string
	) {
		super(message);
		this.name = 'YahooFinanceError';
	}
}

// =============================================================================
// Internal Helpers
// =============================================================================

/**
 * Rate-limited fetch wrapper
 */
async function rateLimitedFetch(url: string): Promise<Response> {
	return yahooRateLimiter.enqueue(async () => {
		const response = await fetch(url, {
			headers: {
				'User-Agent': 'Bagholdr/1.0',
				Accept: 'application/json'
			}
		});
		return response;
	});
}

/**
 * Convert Unix timestamp (seconds) to YYYY-MM-DD string
 */
function timestampToDate(timestamp: number): string {
	const date = new Date(timestamp * 1000);
	return date.toISOString().split('T')[0];
}

// =============================================================================
// Symbol Resolution
// =============================================================================

/**
 * Resolve ISIN to all available Yahoo Finance symbols
 * Returns all exchanges where this ISIN is available
 */
export async function fetchAllSymbolsFromIsin(isin: string): Promise<YahooSearchResult[]> {
	const url = `${YAHOO_SEARCH_URL}?q=${encodeURIComponent(isin)}&newsCount=0&listsCount=0&quotesCount=20&quotesQueryId=tss_match_phrase_query`;

	const response = await rateLimitedFetch(url);

	if (response.status === 403) {
		throw new YahooFinanceError('Rate limited by Yahoo Finance', 'RATE_LIMITED');
	}

	if (!response.ok) {
		throw new YahooFinanceError(
			`HTTP error fetching symbol: ${response.status}`,
			'HTTP_ERROR'
		);
	}

	const data = await response.json();

	if (!data.quotes || data.quotes.length === 0) {
		throw new YahooFinanceError(`No symbol found for ISIN: ${isin}`, 'NOT_FOUND');
	}

	return data.quotes.map((q: Record<string, unknown>) => ({
		symbol: q.symbol as string,
		exchange: q.exchange as string | undefined,
		exchangeDisplay: q.exchDisp as string | undefined,
		quoteType: q.quoteType as string | undefined,
		shortname: q.shortname as string | undefined
	}));
}

/**
 * Resolve ISIN to Yahoo Finance ticker symbol (first/best match)
 */
export async function fetchSymbolFromIsin(isin: string): Promise<string> {
	const symbols = await fetchAllSymbolsFromIsin(isin);
	return symbols[0].symbol;
}

// =============================================================================
// Current Price
// =============================================================================

/**
 * Fetch current price data for a ticker symbol
 */
export async function fetchPriceData(symbol: string): Promise<YahooPriceInfo> {
	const url = `${YAHOO_CHART_URL}/${encodeURIComponent(symbol)}`;

	const response = await rateLimitedFetch(url);

	if (response.status === 403) {
		throw new YahooFinanceError('Rate limited by Yahoo Finance', 'RATE_LIMITED');
	}

	if (!response.ok) {
		throw new YahooFinanceError(
			`HTTP error fetching price: ${response.status}`,
			'HTTP_ERROR'
		);
	}

	const data = await response.json();

	if (!data.chart?.result?.[0]?.meta) {
		throw new YahooFinanceError(`No price data found for: ${symbol}`, 'NOT_FOUND');
	}

	const meta = data.chart.result[0].meta;

	return {
		price: meta.regularMarketPrice,
		currency: meta.currency,
		instrumentType: meta.instrumentType ?? 'unknown',
		timestamp: new Date()
	};
}

// =============================================================================
// Historical Data
// =============================================================================

/**
 * Fetch historical daily price data for a ticker symbol
 *
 * @param symbol - Yahoo Finance ticker symbol
 * @param range - Time range: '1y', '5y', '10y', or 'max' (default: '10y')
 *
 * Note: Yahoo automatically downsamples to monthly data for very long ranges.
 * Using '10y' instead of 'max' ensures we get daily granularity for most instruments.
 * If you need more history, use 'max' but be aware the granularity may be monthly.
 *
 * @returns Historical candles and dividend events
 */
export async function fetchHistoricalData(
	symbol: string,
	range: '1y' | '5y' | '10y' | 'max' = '10y'
): Promise<YahooHistoricalData> {
	// Fetch daily candles with dividend events
	const url = `${YAHOO_CHART_URL}/${encodeURIComponent(symbol)}?interval=1d&range=${range}&events=div`;

	const response = await rateLimitedFetch(url);

	if (response.status === 403) {
		throw new YahooFinanceError('Rate limited by Yahoo Finance', 'RATE_LIMITED');
	}

	if (response.status === 404) {
		throw new YahooFinanceError(`Ticker not found: ${symbol}`, 'NOT_FOUND');
	}

	if (!response.ok) {
		throw new YahooFinanceError(
			`HTTP error fetching historical data: ${response.status}`,
			'HTTP_ERROR'
		);
	}

	const data = await response.json();

	if (!data.chart?.result?.[0]) {
		throw new YahooFinanceError(`No historical data found for: ${symbol}`, 'NOT_FOUND');
	}

	const result = data.chart.result[0];
	const meta = result.meta;

	// Warn if Yahoo returned a different granularity than requested (1d)
	const actualGranularity = meta.dataGranularity;
	if (actualGranularity && actualGranularity !== '1d') {
		console.warn(
			`[Yahoo] ${symbol}: Requested daily data but got ${actualGranularity} granularity. ` +
				`Consider using a shorter range (current: ${range}). Got ${result.timestamp?.length ?? 0} data points.`
		);
	}

	const timestamps: number[] = result.timestamp ?? [];
	const quote = result.indicators?.quote?.[0] ?? {};
	const adjClose = result.indicators?.adjclose?.[0]?.adjclose ?? [];
	const events = result.events ?? {};

	// Parse candles
	const candles: YahooCandle[] = [];
	for (let i = 0; i < timestamps.length; i++) {
		// Skip if any required value is null/undefined
		if (
			quote.open?.[i] == null ||
			quote.high?.[i] == null ||
			quote.low?.[i] == null ||
			quote.close?.[i] == null ||
			adjClose[i] == null
		) {
			continue;
		}

		candles.push({
			date: timestampToDate(timestamps[i]),
			open: quote.open[i],
			high: quote.high[i],
			low: quote.low[i],
			close: quote.close[i],
			adjClose: adjClose[i],
			volume: quote.volume?.[i] ?? 0
		});
	}

	// Parse dividends
	const dividends: YahooDividend[] = [];
	if (events.dividends) {
		for (const [timestamp, div] of Object.entries(events.dividends)) {
			const dividend = div as { amount?: number; date?: number };
			if (dividend.amount != null) {
				dividends.push({
					exDate: timestampToDate(parseInt(timestamp)),
					amount: dividend.amount
				});
			}
		}
	}

	// Sort dividends by date
	dividends.sort((a, b) => a.exDate.localeCompare(b.exDate));

	return {
		candles,
		dividends,
		currency: meta.currency ?? 'USD'
	};
}

/**
 * Fetch intraday price data (5-minute intervals) for a ticker symbol
 *
 * @param symbol - Yahoo Finance ticker symbol
 * @param range - Time range: '1d', '5d' (default: '5d' for 5 days of data)
 * @returns Intraday candles with timestamps
 */
export async function fetchIntradayData(
	symbol: string,
	range: '1d' | '5d' = '5d'
): Promise<YahooIntradayData> {
	// Fetch 5-minute candles
	const url = `${YAHOO_CHART_URL}/${encodeURIComponent(symbol)}?interval=5m&range=${range}&includePrePost=false`;

	const response = await rateLimitedFetch(url);

	if (response.status === 403) {
		throw new YahooFinanceError('Rate limited by Yahoo Finance', 'RATE_LIMITED');
	}

	if (response.status === 404) {
		throw new YahooFinanceError(`Ticker not found: ${symbol}`, 'NOT_FOUND');
	}

	if (!response.ok) {
		throw new YahooFinanceError(
			`HTTP error fetching intraday data: ${response.status}`,
			'HTTP_ERROR'
		);
	}

	const data = await response.json();

	if (!data.chart?.result?.[0]) {
		throw new YahooFinanceError(`No intraday data found for: ${symbol}`, 'NOT_FOUND');
	}

	const result = data.chart.result[0];
	const meta = result.meta;
	const timestamps: number[] = result.timestamp ?? [];
	const quote = result.indicators?.quote?.[0] ?? {};

	// Parse candles
	const candles: YahooIntradayCandle[] = [];
	for (let i = 0; i < timestamps.length; i++) {
		// Skip if any required value is null/undefined
		if (
			quote.open?.[i] == null ||
			quote.high?.[i] == null ||
			quote.low?.[i] == null ||
			quote.close?.[i] == null
		) {
			continue;
		}

		candles.push({
			timestamp: timestamps[i],
			open: quote.open[i],
			high: quote.high[i],
			low: quote.low[i],
			close: quote.close[i],
			volume: quote.volume?.[i] ?? 0
		});
	}

	return {
		candles,
		currency: meta.currency ?? 'USD'
	};
}

// =============================================================================
// FX Rates
// =============================================================================

/**
 * Fetch FX rate from one currency to another
 * Uses Yahoo Finance forex pairs (e.g., USDEUR=X)
 */
export async function fetchFxRate(from: string, to: string): Promise<number> {
	// Normalize currency codes
	const fromUpper = from.toUpperCase();
	const toUpper = to.toUpperCase();

	// If same currency, rate is 1
	if (fromUpper === toUpper) {
		return 1;
	}

	// Handle GBp (pence) -> GBP conversion
	const actualFrom = fromUpper === 'GBP' ? 'GBP' : fromUpper;

	const symbol = `${actualFrom}${toUpper}=X`;
	const priceInfo = await fetchPriceData(symbol);

	let rate = priceInfo.price;

	// Special handling for GBp (British pence)
	// Yahoo returns the rate for GBP, but prices might be in pence
	if (from.toLowerCase() === 'gbp' && from !== 'GBP') {
		// Original was lowercase 'gbp' meaning pence
		rate = rate / 100;
	}

	return rate;
}

/**
 * Adjust conversion rate for special cases like GBp (pence)
 */
export function adjustConversionRate(currency: string, rate: number): number {
	// GBp (pence) - prices are in pence but FX is for pounds
	if (currency === 'GBp') {
		return rate / 100;
	}
	return rate;
}

// =============================================================================
// Convenience Functions
// =============================================================================

/**
 * Fetch price in EUR for an asset
 * Handles ISIN resolution and FX conversion
 */
export async function fetchPriceInEur(isin: string, knownTicker?: string): Promise<{
	ticker: string;
	priceNative: number;
	currency: string;
	priceEur: number;
	instrumentType: string;
}> {
	// Resolve ISIN to ticker if not provided
	const ticker = knownTicker ?? (await fetchSymbolFromIsin(isin));

	// Fetch price in native currency
	const priceInfo = await fetchPriceData(ticker);

	// Convert to EUR if needed
	let priceEur: number;
	if (priceInfo.currency === 'EUR') {
		priceEur = priceInfo.price;
	} else {
		const fxRate = await fetchFxRate(priceInfo.currency, 'EUR');
		const adjustedRate = adjustConversionRate(priceInfo.currency, fxRate);
		priceEur = priceInfo.price * adjustedRate;
	}

	return {
		ticker,
		priceNative: priceInfo.price,
		currency: priceInfo.currency,
		priceEur,
		instrumentType: priceInfo.instrumentType
	};
}

File: native/bagholdr/bagholdr_server/lib/src/oracle/yahoo.dart
import 'dart:convert';
import 'dart:io';

import 'rate_limiter.dart';

// =============================================================================
// Constants
// =============================================================================

const String yahooSearchUrl =
    'https://query1.finance.yahoo.com/v1/finance/search';
const String yahooChartUrl =
    'https://query1.finance.yahoo.com/v8/finance/chart';

// =============================================================================
// Types
// =============================================================================

class YahooPriceInfo {
  final double price;
  final String currency;
  final String instrumentType;
  final DateTime timestamp;

  YahooPriceInfo({
    required this.price,
    required this.currency,
    required this.instrumentType,
    required this.timestamp,
  });
}

class YahooSearchResult {
  final String symbol;
  final String? exchange;
  final String? exchangeDisplay;
  final String? quoteType;
  final String? shortname;

  YahooSearchResult({
    required this.symbol,
    this.exchange,
    this.exchangeDisplay,
    this.quoteType,
    this.shortname,
  });
}

class YahooCandle {
  final String date; // YYYY-MM-DD
  final double open;
  final double high;
  final double low;
  final double close;
  final double adjClose;
  final int volume;

  YahooCandle({
    required this.date,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.adjClose,
    required this.volume,
  });
}

class YahooDividend {
  final String exDate; // YYYY-MM-DD
  final double amount;

  YahooDividend({required this.exDate, required this.amount});
}

class YahooHistoricalData {
  final List<YahooCandle> candles;
  final List<YahooDividend> dividends;
  final String currency;

  YahooHistoricalData({
    required this.candles,
    required this.dividends,
    required this.currency,
  });
}

class YahooIntradayCandle {
  final int timestamp; // Unix timestamp in seconds
  final double open;
  final double high;
  final double low;
  final double close;
  final int volume;

  YahooIntradayCandle({
    required this.timestamp,
    required this.open,
    required this.high,
    required this.low,
    required this.close,
    required this.volume,
  });
}

class YahooIntradayData {
  final List<YahooIntradayCandle> candles;
  final String currency;

  YahooIntradayData({required this.candles, required this.currency});
}

class YahooFinanceError implements Exception {
  final String message;
  final String code;

  YahooFinanceError(this.message, this.code);

  @override
  String toString() => 'YahooFinanceError($code): $message';
}

// =============================================================================
// Internal Helpers
// =============================================================================

final _httpClient = HttpClient()
  ..userAgent = 'Bagholdr/1.0';

/// Rate-limited fetch wrapper. Returns decoded JSON.
Future<Map<String, dynamic>> _rateLimitedFetch(String url) {
  return yahooRateLimiter.enqueue(() async {
    final uri = Uri.parse(url);
    final request = await _httpClient.getUrl(uri);
    request.headers.set('Accept', 'application/json');
    final response = await request.close();

    if (response.statusCode == 403) {
      await response.drain<void>();
      throw YahooFinanceError(
        'Rate limited by Yahoo Finance',
        'RATE_LIMITED',
      );
    }

    if (response.statusCode == 404) {
      await response.drain<void>();
      throw YahooFinanceError(
        'Not found: $url',
        'NOT_FOUND',
      );
    }

    if (response.statusCode != 200) {
      await response.drain<void>();
      throw YahooFinanceError(
        'HTTP error: ${response.statusCode}',
        'HTTP_ERROR',
      );
    }

    final body = await response.transform(utf8.decoder).join();
    return jsonDecode(body) as Map<String, dynamic>;
  });
}

/// Convert Unix timestamp (seconds) to YYYY-MM-DD string.
String _timestampToDate(int timestamp) {
  final date = DateTime.fromMillisecondsSinceEpoch(timestamp * 1000, isUtc: true);
  return '${date.year.toString().padLeft(4, '0')}-'
      '${date.month.toString().padLeft(2, '0')}-'
      '${date.day.toString().padLeft(2, '0')}';
}

// =============================================================================
// Symbol Resolution
// =============================================================================

/// Resolve ISIN to all available Yahoo Finance symbols.
/// Returns all exchanges where this ISIN is available.
Future<List<YahooSearchResult>> fetchAllSymbolsFromIsin(String isin) async {
  final url = '$yahooSearchUrl?q=${Uri.encodeComponent(isin)}'
      '&newsCount=0&listsCount=0&quotesCount=20'
      '&quotesQueryId=tss_match_phrase_query';

  final data = await _rateLimitedFetch(url);

  final quotes = data['quotes'] as List<dynamic>?;
  if (quotes == null || quotes.isEmpty) {
    throw YahooFinanceError('No symbol found for ISIN: $isin', 'NOT_FOUND');
  }

  return quotes.map((q) {
    final map = q as Map<String, dynamic>;
    return YahooSearchResult(
      symbol: map['symbol'] as String,
      exchange: map['exchange'] as String?,
      exchangeDisplay: map['exchDisp'] as String?,
      quoteType: map['quoteType'] as String?,
      shortname: map['shortname'] as String?,
    );
  }).toList();
}

/// Resolve ISIN to Yahoo Finance ticker symbol (first/best match).
Future<String> fetchSymbolFromIsin(String isin) async {
  final symbols = await fetchAllSymbolsFromIsin(isin);
  return symbols.first.symbol;
}

// =============================================================================
// Current Price
// =============================================================================

/// Fetch current price data for a ticker symbol.
Future<YahooPriceInfo> fetchPriceData(String symbol) async {
  final url = '$yahooChartUrl/${Uri.encodeComponent(symbol)}';

  final data = await _rateLimitedFetch(url);

  final chart = data['chart'] as Map<String, dynamic>?;
  final results = chart?['result'] as List<dynamic>?;
  if (results == null || results.isEmpty) {
    throw YahooFinanceError('No price data found for: $symbol', 'NOT_FOUND');
  }

  final result = results[0] as Map<String, dynamic>;
  final meta = result['meta'] as Map<String, dynamic>?;
  if (meta == null) {
    throw YahooFinanceError('No price data found for: $symbol', 'NOT_FOUND');
  }

  return YahooPriceInfo(
    price: (meta['regularMarketPrice'] as num).toDouble(),
    currency: meta['currency'] as String? ?? 'USD',
    instrumentType: meta['instrumentType'] as String? ?? 'unknown',
    timestamp: DateTime.now(),
  );
}

// =============================================================================
// Historical Data
// =============================================================================

/// Fetch historical daily price data for a ticker symbol.
///
/// [range] - Time range: '1y', '5y', '10y', or 'max' (default: '10y')
///
/// Note: Yahoo automatically downsamples to monthly data for very long ranges.
/// Using '10y' instead of 'max' ensures we get daily granularity for most instruments.
Future<YahooHistoricalData> fetchHistoricalData(
  String symbol, {
  String range = '10y',
}) async {
  final url = '$yahooChartUrl/${Uri.encodeComponent(symbol)}'
      '?interval=1d&range=$range&events=div';

  final data = await _rateLimitedFetch(url);

  final chart = data['chart'] as Map<String, dynamic>?;
  final results = chart?['result'] as List<dynamic>?;
  if (results == null || results.isEmpty) {
    throw YahooFinanceError(
      'No historical data found for: $symbol',
      'NOT_FOUND',
    );
  }

  final result = results[0] as Map<String, dynamic>;
  final meta = result['meta'] as Map<String, dynamic>? ?? {};
  final timestamps = (result['timestamp'] as List<dynamic>?)
          ?.map((t) => (t as num).toInt())
          .toList() ??
      [];

  final indicators = result['indicators'] as Map<String, dynamic>? ?? {};
  final quoteList = indicators['quote'] as List<dynamic>?;
  final quote = quoteList != null && quoteList.isNotEmpty
      ? quoteList[0] as Map<String, dynamic>
      : <String, dynamic>{};
  final adjCloseList = indicators['adjclose'] as List<dynamic>?;
  final adjCloseArr = adjCloseList != null && adjCloseList.isNotEmpty
      ? (adjCloseList[0] as Map<String, dynamic>)['adjclose'] as List<dynamic>?
      : null;

  final events = result['events'] as Map<String, dynamic>? ?? {};

  // Parse candles
  final candles = <YahooCandle>[];
  final openArr = quote['open'] as List<dynamic>?;
  final highArr = quote['high'] as List<dynamic>?;
  final lowArr = quote['low'] as List<dynamic>?;
  final closeArr = quote['close'] as List<dynamic>?;
  final volumeArr = quote['volume'] as List<dynamic>?;

  for (var i = 0; i < timestamps.length; i++) {
    final open = openArr?[i];
    final high = highArr?[i];
    final low = lowArr?[i];
    final close = closeArr?[i];
    final adjClose = adjCloseArr?[i];

    // Skip if any required value is null
    if (open == null || high == null || low == null || close == null || adjClose == null) {
      continue;
    }

    candles.add(YahooCandle(
      date: _timestampToDate(timestamps[i]),
      open: (open as num).toDouble(),
      high: (high as num).toDouble(),
      low: (low as num).toDouble(),
      close: (close as num).toDouble(),
      adjClose: (adjClose as num).toDouble(),
      volume: (volumeArr?[i] as num?)?.toInt() ?? 0,
    ));
  }

  // Parse dividends
  final dividends = <YahooDividend>[];
  final dividendEvents = events['dividends'] as Map<String, dynamic>?;
  if (dividendEvents != null) {
    for (final entry in dividendEvents.entries) {
      final div = entry.value as Map<String, dynamic>;
      final amount = div['amount'] as num?;
      if (amount != null) {
        dividends.add(YahooDividend(
          exDate: _timestampToDate(int.parse(entry.key)),
          amount: amount.toDouble(),
        ));
      }
    }
  }

  // Sort dividends by date
  dividends.sort((a, b) => a.exDate.compareTo(b.exDate));

  return YahooHistoricalData(
    candles: candles,
    dividends: dividends,
    currency: meta['currency'] as String? ?? 'USD',
  );
}

/// Fetch intraday price data (5-minute intervals) for a ticker symbol.
///
/// [range] - Time range: '1d', '5d' (default: '5d' for 5 days of data)
Future<YahooIntradayData> fetchIntradayData(
  String symbol, {
  String range = '5d',
}) async {
  final url = '$yahooChartUrl/${Uri.encodeComponent(symbol)}'
      '?interval=5m&range=$range&includePrePost=false';

  final data = await _rateLimitedFetch(url);

  final chart = data['chart'] as Map<String, dynamic>?;
  final results = chart?['result'] as List<dynamic>?;
  if (results == null || results.isEmpty) {
    throw YahooFinanceError(
      'No intraday data found for: $symbol',
      'NOT_FOUND',
    );
  }

  final result = results[0] as Map<String, dynamic>;
  final meta = result['meta'] as Map<String, dynamic>? ?? {};
  final timestamps = (result['timestamp'] as List<dynamic>?)
          ?.map((t) => (t as num).toInt())
          .toList() ??
      [];

  final indicators = result['indicators'] as Map<String, dynamic>? ?? {};
  final quoteList = indicators['quote'] as List<dynamic>?;
  final quote = quoteList != null && quoteList.isNotEmpty
      ? quoteList[0] as Map<String, dynamic>
      : <String, dynamic>{};

  // Parse candles
  final candles = <YahooIntradayCandle>[];
  final openArr = quote['open'] as List<dynamic>?;
  final highArr = quote['high'] as List<dynamic>?;
  final lowArr = quote['low'] as List<dynamic>?;
  final closeArr = quote['close'] as List<dynamic>?;
  final volumeArr = quote['volume'] as List<dynamic>?;

  for (var i = 0; i < timestamps.length; i++) {
    final open = openArr?[i];
    final high = highArr?[i];
    final low = lowArr?[i];
    final close = closeArr?[i];

    if (open == null || high == null || low == null || close == null) {
      continue;
    }

    candles.add(YahooIntradayCandle(
      timestamp: timestamps[i],
      open: (open as num).toDouble(),
      high: (high as num).toDouble(),
      low: (low as num).toDouble(),
      close: (close as num).toDouble(),
      volume: (volumeArr?[i] as num?)?.toInt() ?? 0,
    ));
  }

  return YahooIntradayData(
    candles: candles,
    currency: meta['currency'] as String? ?? 'USD',
  );
}

// =============================================================================
// FX Rates
// =============================================================================

/// Fetch FX rate from one currency to another.
/// Uses Yahoo Finance forex pairs (e.g., USDEUR=X).
Future<double> fetchFxRate(String from, String to) async {
  final fromUpper = from.toUpperCase();
  final toUpper = to.toUpperCase();

  // Same currency
  if (fromUpper == toUpper) return 1.0;

  // Handle GBp (pence) -> GBP conversion
  final actualFrom = fromUpper == 'GBP' ? 'GBP' : fromUpper;

  final symbol = '$actualFrom$toUpper=X';
  final priceInfo = await fetchPriceData(symbol);

  var rate = priceInfo.price;

  // Special handling for GBp (British pence)
  // Yahoo returns the rate for GBP, but prices might be in pence
  if (from.toLowerCase() == 'gbp' && from != 'GBP') {
    rate = rate / 100;
  }

  return rate;
}

/// Adjust conversion rate for special cases like GBp (pence).
double adjustConversionRate(String currency, double rate) {
  if (currency == 'GBp') {
    return rate / 100;
  }
  return rate;
}

// =============================================================================
// Convenience Functions
// =============================================================================

class PriceInEurResult {
  final String ticker;
  final double priceNative;
  final String currency;
  final double priceEur;
  final String instrumentType;

  PriceInEurResult({
    required this.ticker,
    required this.priceNative,
    required this.currency,
    required this.priceEur,
    required this.instrumentType,
  });
}

/// Fetch price in EUR for an asset.
/// Handles ISIN resolution and FX conversion.
Future<PriceInEurResult> fetchPriceInEur(
  String isin, {
  String? knownTicker,
}) async {
  // Resolve ISIN to ticker if not provided
  final ticker = knownTicker ?? await fetchSymbolFromIsin(isin);

  // Fetch price in native currency
  final priceInfo = await fetchPriceData(ticker);

  // Convert to EUR if needed
  double priceEur;
  if (priceInfo.currency == 'EUR') {
    priceEur = priceInfo.price;
  } else {
    final fxRate = await fetchFxRate(priceInfo.currency, 'EUR');
    final adjustedRate = adjustConversionRate(priceInfo.currency, fxRate);
    priceEur = priceInfo.price * adjustedRate;
  }

  return PriceInEurResult(
    ticker: ticker,
    priceNative: priceInfo.price,
    currency: priceInfo.currency,
    priceEur: priceEur,
    instrumentType: priceInfo.instrumentType,
  );
}
